### PR TITLE
Handle Spaces In Tab Completion Logic

### PR DESCRIPTION
**Pull request by zeroSteiner**
This Pull Request adjusts how strings are split on the command line as part of the tab completion logic to support escaped paths with spaces in them. More specifically, it switches the split logic to use the `Shellwords.split` function in place of a regex which simply split on white space to more closely align the completion parsing with how the command itself parsed. This has the notable benefit of allowing the user to tab complete file paths which contain spaces.

If the user ends a line with `\ ` (or any odd number of `\`), it will include the space at the end of the last word for the purposes of completion.

## Example Usage

1. Create the following directory structure
    * `/tmp/Metasploit Test/some target list.txt`
    * `/tmp/Metasploit Test/some_other_targets.txt`
    * `/tmp/Metasploit Test/Another Directory/more targets.txt`
1. Start `./msfconsole`
1. Use a module that takes a file as an option (pro-tip all `RHOSTS` can using the `file:` syntax)
    * [ ] `use exploit/windows/smb/psexec`
    * [ ] Type `set RHOSTS file:/tmp/Metasploit<TAB>`
        * [ ] See it complete `set RHOSTS /tmp/Metasploit\ Test/`
    * [ ] Type `set RHOSTS file:/tmp/Metasploit\ Test/<TAB><TAB>`
        * [ ] See it suggest `Another\ Directory      some\ target\ list.txt  some_other_targets.txt`
    * [ ] Type `set RHOSTS file:/tmp/Metasploit\ Test/some\ <TAB>`
        * [ ] See it complete `set RHOSTS file:/tmp/Metasploit\ Test/some\ target\ list.txt`

### CODE REVIEW

File: lib/rex/ui/text/dispatcher_shell.rb [hunk 1]
```diff
@@ -359,11 +359,11 @@ def initialize(prompt, prompt_char = '>', histfile = nil, framework = nil)
   #
   def tab_complete(str)
     # Check trailing whitespace so we can tell 'x' from 'x '
-    str_match = str.match(/\s+$/)
+    str_match = str.match(/[^\\]([\\]{2})*\s+$/)
     str_trail = (str_match.nil?) ? '' : str_match[0]
 
     # Split the line up by whitespace into words
-    str_words = str.split(/[\s\t\n]+/)
+    str_words = Shellwords.split(str)
```
Comment (wvu-r7) [inline lib/rex/ui/text/dispatcher_shell.rb]: I'm glad to see more `Shellwords` in our bespoke console, but this is more sensitive to input than `split` and will trigger an `ArgumentError` on unbalanced quotes, for instance.

```
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > options

Module options (exploit/unix/webapp/drupal_drupalgeddon2):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   DUMP_OUTPUT  false            no        Dump payload command output
   PHP_FUNC     passthru         yes       PHP function to execute
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                        yes       The target address range or CIDR identifier
   RPORT        80               yes       The target port (TCP)
   SSL          false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /                yes       Path to Drupal install
   VHOST                         no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   Automatic (PHP In-Memory)


msf5 exploit(unix/webapp/drupal_drupalgeddon2) > show targets

Exploit targets:

   Id  Name
   --  ----
   0   Automatic (PHP In-Memory)
   1   Automatic (PHP Dropper)
   2   Automatic (Unix In-Memory)
   3   Automatic (Linux Dropper)
   4   Drupal 7.x (PHP In-Memory)
   5   Drupal 7.x (PHP Dropper)
   6   Drupal 7.x (Unix In-Memory)
   7   Drupal 7.x (Linux Dropper)
   8   Drupal 8.x (PHP In-Memory)
   9   Drupal 8.x (PHP Dropper)
   10  Drupal 8.x (Unix In-Memory)
   11  Drupal 8.x (Linux Dropper)


msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target Automatic\ (PHP\ Dropper)
target => Automatic (PHP Dropper)
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > options

Module options (exploit/unix/webapp/drupal_drupalgeddon2):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   DUMP_OUTPUT  false            no        Dump payload command output
   PHP_FUNC     passthru         yes       PHP function to execute
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                        yes       The target address range or CIDR identifier
   RPORT        80               yes       The target port (TCP)
   SSL          false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /                yes       Path to Drupal install
   VHOST                         no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   1   Automatic (PHP Dropper)


msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "AuTraceback (most recent call last):
	20: from ./msfconsole:49:in `<main>'
	19: from /rapid7/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
	18: from /rapid7/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
	17: from /rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:134:in `run'
	16: from /rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:309:in `get_input_line'
	15: from /rapid7/metasploit-framework/lib/rex/ui/text/input/readline.rb:100:in `pgets'
	14: from /rapid7/metasploit-framework/lib/rex/ui/text/input/readline.rb:162:in `readline_with_output'
	13: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4875:in `readline'
	12: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4853:in `readline_internal'
	11: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4779:in `readline_internal_charloop'
	10: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4363:in `_rl_dispatch'
	 9: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4374:in `_rl_dispatch_subseq'
	 8: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6903:in `rl_complete'
	 7: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6813:in `rl_complete_internal'
	 6: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6329:in `gen_completion_matches'
	 5: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/readline.rb:136:in `readline_attempted_completion_function'
	 4: from /rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:62:in `block in init_tab_complete'
	 3: from /rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:366:in `tab_complete'
	 2: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/2.6.0/shellwords.rb:82:in `shellsplit'
	 1: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/2.6.0/shellwords.rb:82:in `scan'
/Users/wvu/.rbenv/versions/2.6.2/lib/ruby/2.6.0/shellwords.rb:84:in `block in shellsplit': Unmatched double quote: "set target \"Au" (ArgumentError)
wvu@kharak:/rapid7/metasploit-framework:HEAD$
```

**Review by wvu — CHANGES_REQUESTED** [review]
Hi, @zeroSteiner! Great work on this! You've fixed a lot of the unsatisfied expectations people have had about our tab completion. For one, spaces are handled appropriately now!

**Comment by wvu-r7** [comment]
This PR addresses the design limitation in word splitting referenced in #9962.

File: lib/rex/ui/text/dispatcher_shell.rb (hunk 1 above)
Comment (zeroSteiner) [reply lib/rex/ui/text/dispatcher_shell.rb]: Whoops yeah we definitely can't have any of that. I'll figure out how to fix this and push a new version. Great catch @wvu-r7 :+1: 

Are there other instances you can think of where an `ArgumentError` would be raised? I'll look through the `Shellwords` docs of course, but if you know of something else I'll be sure to handle it.

Comment (wvu-r7) [reply lib/rex/ui/text/dispatcher_shell.rb]: I think it's only unmatched quotes. It's not just double quotes, so the error message is a lie.

https://github.com/ruby/ruby/blob/trunk/lib/shellwords.rb
https://github.com/ruby/ruby/blob/trunk/test/test_shellwords.rb

**Comment by timwr** [comment]
Great work! I noticed that the `meterpreter > download/cat/cd (remote file)` tab completion always appends an unnecessary space after the result. Fingers crossed that is fixed too (otherwise I'll have another look).

**Comment by wvu-r7** [comment]
Lol, I want #10850 to be fixed, but I haven't checked with this PR yet.

File: lib/rex/ui/text/dispatcher_shell.rb (hunk 1 above)
Comment (zeroSteiner) [reply lib/rex/ui/text/dispatcher_shell.rb]: So I implemented a custom `shellsplit` function, which when unbalanced quotes are present, will treat the rest of the string as if the last character were the closing quote. Both the words and the unbalanced quote (if present) are then returned in a hash. The quote then get's passed through to `tab_complete_stub` where it's prepended to words and the words do not have their spaces escaped.

Should work quite nicely now for your `set TARGET "Au` example.

Comment (wvu-r7) [reply lib/rex/ui/text/dispatcher_shell.rb]: Wow!!

File: lib/rex/ui/text/dispatcher_shell.rb [hunk 2]
```diff
@@ -621,6 +622,36 @@ def unblock_command(cmd)
     self.blocked.delete(cmd)
   end
 
+  #
+  # Split a line as Shellwords.split would however instead of raising an
+  # ArgumentError on unbalanced quotes return the remainder of the string as if
+  # the last character were the closing quote.
+  #
+  def shellsplitex(line)
```
Comment (wvu-r7) [inline lib/rex/ui/text/dispatcher_shell.rb]: This method is magic sauce.

**Comment by wvu-r7** [comment]
So the behavior of the new `shellsplitex` is a little weird on a starting quote:

```
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set "target
```

It works fine when I start the string.

```
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "Automatic (
set target "Automatic (Linux Dropper)"   set target "Automatic (PHP Dropper)"     set target "Automatic (PHP In-Memory)"   set target "Automatic (Unix In-Memory)"
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "Automatic (
```

Behavior in `bash`:

```
wvu@kharak:/rapid7/metasploit-framework:HEAD$ cd "
.git/          app/           data/          docker/        external/      modules/       script/        spec/          tools/
.github/       config/        db/            documentation/ lib/           plugins/       scripts/       test/
wvu@kharak:/rapid7/metasploit-framework:HEAD$ cd "
```

It's probably just off by one?

**Comment by zeroSteiner** [comment]
Okay, 3rd times the charm. It should be appending an empty string to the end of words where the last character is an unbalanced quote which fixes the issue. See commit dfa37a9fb094ad9bdc177c223861383190ed4f57.

Thanks for testing this so thoroughly and bearing with me here.

**Comment by wvu-r7** [comment]
It's all good. We have a very involved implementation. You took this on when no one else tried!

**Comment by wvu-r7** [comment]
Perfect.

```
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target
set target     set targeturi
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target
set target 0                               set target 3                               set target 8                               set target Automatic\ (Unix\ In-Memory)    set target Drupal\ 8.x\ (Linux\ Dropper)
set target 1                               set target 4                               set target 9                               set target Drupal\ 7.x\ (Linux\ Dropper)   set target Drupal\ 8.x\ (PHP\ Dropper)
set target 10                              set target 5                               set target Automatic\ (Linux\ Dropper)     set target Drupal\ 7.x\ (PHP\ Dropper)     set target Drupal\ 8.x\ (PHP\ In-Memory)
set target 11                              set target 6                               set target Automatic\ (PHP\ Dropper)       set target Drupal\ 7.x\ (PHP\ In-Memory)   set target Drupal\ 8.x\ (Unix\ In-Memory)
set target 2                               set target 7                               set target Automatic\ (PHP\ In-Memory)     set target Drupal\ 7.x\ (Unix\ In-Memory)
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "
set target "0"                            set target "3"                            set target "8"                            set target "Automatic (Unix In-Memory)"   set target "Drupal 8.x (Linux Dropper)"
set target "1"                            set target "4"                            set target "9"                            set target "Drupal 7.x (Linux Dropper)"   set target "Drupal 8.x (PHP Dropper)"
set target "10"                           set target "5"                            set target "Automatic (Linux Dropper)"    set target "Drupal 7.x (PHP Dropper)"     set target "Drupal 8.x (PHP In-Memory)"
set target "11"                           set target "6"                            set target "Automatic (PHP Dropper)"      set target "Drupal 7.x (PHP In-Memory)"   set target "Drupal 8.x (Unix In-Memory)"
set target "2"                            set target "7"                            set target "Automatic (PHP In-Memory)"    set target "Drupal 7.x (Unix In-Memory)"
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "Automatic (Linux Dropper)"
target => Automatic (Linux Dropper)
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target Automatic\ (PHP\ Dropper)
target => Automatic (PHP Dropper)
msf5 exploit(unix/webapp/drupal_drupalgeddon2) >
```

This might just be my favorite core enhancement this year.

**Review by wvu — APPROVED** [review]

**Comment by wvu-r7** [comment]
# Release Notes

This fixes a long-standing flaw in `msfconsole`'s tab completion implementation where any line containing spaces would fail to complete. Both quoting and escaping are offered as solutions to the problem of word splitting.

**Pull request merged**